synapse/rest/media/v1/preview_url_resource.py

```diff
@@ -417,6 +418,38 @@ async def _download_url(
 
         return length, uri, code, media_type, download_name, expires, etag
 
+    async def _parse_data_url(
+        self, url: str, output_stream: BinaryIO
+    ) -> Tuple[int, str, int, str, Optional[str], int, Optional[str]]:
+        """
+        Parses a data: URL and stores it.
+        """
+        try:
+            logger.debug("Trying to parse data url '%s'", url)
+            with urlopen(url) as url_info:
+                # TODO Can this be more efficient.
+                output_stream.write(url_info.read())
+        except Exception as e:
+            logger.warning("Error parsing data: URL %s: %r", url, e)
+
+            raise SynapseError(
+                500,
+                "Failed to parse data URL: %s"
+                % (traceback.format_exception_only(sys.exc_info()[0], e),),
+                Codes.UNKNOWN,
+            )
+
+        length = output_stream.seek(1)
+
+        media_type = url_info.headers.get_content_type()
```

Comment:
Should we cross reference this with the mime-type in the data url?

(Do we want to have some kind of whitelist of trusted mime types which are okay to preview?)

Reply:
This *is* the mime-type from the data URL, it gets parsed out for us and shoved into a `headers` property for some reason.

[The docs for `urlopen`](https://docs.python.org/3/library/urllib.request.html#urllib.request.urlopen) have some info on how `data` URLs are handled:

> For FTP, file, and data URLs [...], this function returns a [`urllib.response.addinfourl`](https://docs.python.org/3/library/urllib.request.html#urllib.response.addinfourl) object.

The `addinfourl` object has a `headers` property:

> Returns the headers of the response in the form of an [`EmailMessage`](https://docs.python.org/3/library/email.message.html#email.message.EmailMessage.get_content_type) instance.

The `EmailMessage` object has a [`get_content_type()`](https://docs.python.org/3/library/email.message.html#email.message.EmailMessage.get_content_type) property:

> Return the message’s content type, coerced to lower case of the form *maintype/subtype*. If there is no *Content-Type* header in the message return the value returned by [`get_default_type()`](https://docs.python.org/3/library/email.message.html#email.message.EmailMessage.get_default_type). If the *Content-Type* header is invalid, return `text/plain`.

This all doesn't really make sense for `data` URLs, but from trial and error it seems to do the correct thing.

Reply:
I added a clarifying comment in 116015f0315f65908f296a010c9f29a767556a8c -- hopefully that helps?